content/chainguard/migration/migrating-python.md

```diff
@@ -0,0 +1,124 @@
+
+## Migrating a Python Application
+
+When migrating most containerized Python applications, we recommend building a virtual environment with any needed Python packages using our provided development images, then copying over the virtual environment to our stripped-down production image. Chainguard Academy hosts [detailed instructions for a multi-stage build for a CLI-based Python script](/chainguard/chainguard-images/getting-started/python). 
+
+The below Dockerfile provides an example of such a multi-stage build for a simple Flask application. You can view a version of this Dockerfile with included sample Flask application and requirements.txt in [this repository](https://github.com/chainguard-dev/cg-images-python-migration/tree/python-only), and the original unmigrated application in the [v0 branch](https://github.com/chainguard-dev/cg-images-python-migration/tree/v0). A more complex setup with reverse proxy orchestrated with Docker Compose is provided in the next section.
+
+```Dockerfile
+# syntax=docker/dockerfile:1
+
+FROM cgr.dev/chainguard/python:latest-dev as dev
+
+WORKDIR /flask-app
+
+RUN python -m venv venv
+ENV PATH="/flask-app/venv/bin":$PATH
+COPY requirements.txt requirements.txt
+RUN pip install -r requirements.txt
+
+FROM cgr.dev/chainguard/python:latest
+
+WORKDIR /flask-app
+
+COPY app.py app.py
+COPY --from=dev /flask-app/venv /flask-app/venv
+ENV PATH="/flask-app/venv/bin:$PATH"
+
+EXPOSE 8000
+
+ENTRYPOINT ["python", "-m", "gunicorn", "-b", "0.0.0.0:8000", "app:app"]
+```
+
+When running an application containerized with the above Dockerfile, the application should be visible on `0.0.0.0:8000`.
+
+As you can see, the primary difference in this Flask application compared to the pre-migration application is the use of a multistage build. In the initial stage, we copy our requirements into the development version of the Python Chainguard Image, initialize a virtual environment, and install needed packages with pip. In the second stage, we copy the virtual environment from the development image, copy the application from the host, set exposed port ;metadata, and run the application with the [Gunicorn](https://gunicorn.org/) WSGI server.
+
+By default, the entrypoint for the Python Chainguard Image is `python` rather than `bash`. However, if you replace the included Python binary with the virtual environment binary on the path as we do above, you should set the entrypoint explicitly, otherwise you will not have access to the packages included in your virtual environment.
+
+We recommend that you pin dependencies to specific versions in your own application. The example Flask application script linked above also enables debug mode, which should be turned off in a production scenario.
+
+You may wish to include the following environmental variables in your Dockerfile. They prevent the buffering of output and prevent the creation of cached bytecode, respectively, and can be included in both stages of the multi-stage build.
```

Comment:
I think these have an extremely marginal benefit, but they're actively recommended so I don't want to be remiss. This is the new language:

You may wish to include the following environmental variables in your Dockerfile. The first prevents the buffering of output, meaning that all messages printed to standard output are immediately printed rather than being held in a cache. The second prevents the creation of cached bytecode, which can marginally reduce image size.